### PR TITLE
Quic ietf 4967 v8

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4967
https://redmine.openinfosecfoundation.org/issues/5143
https://redmine.openinfosecfoundation.org/issues/5166

Describe changes:
- Parses standardized QUIC IETF v1
- bump up rust carte tls_parser already used by rep, so that it can be used by quic
- Rustfmt previous code

suricata-verify-pr: 780
https://github.com/OISF/suricata-verify/pull/780

Replaces https://github.com/OISF/suricata/pull/7122 with naming the keyword `ja3s` with an s if it is from the server

Still to do more generally about quic : see https://redmine.openinfosecfoundation.org/issues/4966 (frame support)

### CODE REVIEW

**Comment by victorjulien** [comment]
`ja3s` is still logged as `ja3`. For a comparison, see our `tls-ja3s` SV test:
```json
{
  "timestamp": "2019-05-15T08:11:18.955582+0000",
  "flow_id": 143691514005172,
  "pcap_cnt": 11,
  "event_type": "tls",
  "src_ip": "2a03:b0c0:0002:00d0:0000:0000:0bd3:4001",
  "src_port": 48106,
  "dest_ip": "2606:2800:0220:0001:0248:1893:25c8:1946",
  "dest_port": 443,
  "proto": "TCP",
  "tls": {
    "subject": "C=US, ST=California, L=Los Angeles, O=Internet Corporation for Assigned Names and Numbers, OU=Technology, CN=www.example.org",
    "issuerdn": "C=US, O=DigiCert Inc, CN=DigiCert SHA2 Secure Server CA",
    "serial": "0F:D0:78:DD:48:F1:A2:BD:4D:0F:2B:A9:6B:60:38:FE",
    "fingerprint": "7b:b6:98:38:69:70:36:3d:29:19:cc:57:72:84:69:84:ff:d4:a8:89",
    "sni": "example.com",
    "version": "TLS 1.2",
    "notbefore": "2018-11-28T00:00:00",
    "notafter": "2020-12-02T12:00:00",
    "ja3": {
      "hash": "1fe4c7a3544eb27afec2adfb3a3dbf60",
      "string": "771,49196-49200-159-52393-52392-52394-49195-49199-158-49188-49192-107-49187-49191-103-49162-49172-57-49161-49171-51-157-156-61-60-53-47-255,0-11-10-13172-16-22-23-13,29-23-25-24,0-1-2"
    },
    "ja3s": {
      "hash": "5d79edf64e03689ff559a54e9d9487bc",
      "string": "771,49199,65281-0-11-16-23"
    }
  }
}
```
As this is a bidir tx it has both `ja3` (client) and `ja3s` (server). In the quic case we just need to make sure the fields are named properly, so a tx will have either `ja3` or `ja3s`.

Can we also log both the hash and the string link in `tls`?

**Comment by victorjulien** [comment]
Can you add `quic-events.rules`? I wanted to see why I have so many errors:
```
app_layer.flow.quic                           | Total                     | 102201
app_layer.tx.quic                             | Total                     | 145479
app_layer.error.quic.parser                   | Total                     | 5277807
```

**Comment by victorjulien** [comment]
Btw eve has no `anomaly` records for these errors either. So we may not be setting events?

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | test | baseline | % |
|--|--|--|--| 
| | tlpr1_stats_chk | | |
| .app_layer.flow.quic | 33523 | 39556 | 84.75% |
| .app_layer.tx.quic | 106109 | 4198 | 2527.61% |
| .app_layer.error.quic.parser | 54065 | 43951 | 123.01% | 

Pipeline 6538

**Comment by victorjulien** [comment]
> WARNING:
> 
> field	test	baseline	%
> tlpr1_stats_chk		
> .app_layer.flow.quic	33523	39556	84.75%
> .app_layer.tx.quic	106109	4198	2527.61%
> .app_layer.error.quic.parser	54065	43951	123.01%
> Pipeline 6538

Fewer quic flows... is that expected?

File: rust/src/quic/frames.rs [hunk 1]
```diff
@@ -115,13 +122,184 @@ impl fmt::Display for StreamTag {
...
+            _ => {}
+        }
+    }
+}
+
+// get interesting stuff out of parsed tls extensions
+fn quic_get_tls_extensions(
+    input: Option<&[u8]>, ja3: &mut String, client: bool,
+) -> Vec<QuicTlsExtension> {
+    let mut extv = Vec::new();
```
Comment (jasonish) [inline rust/src/quic/frames.rs:220]: How often, if ever would we endup not adding anything to this vector?

**Comment by catenacyber** [comment]
> Can we also log both the hash and the string link in tls?

Done in next version

I did not do the detection part, as it is only applying a to_md5 transform...

**Comment by catenacyber** [comment]
> Can you add quic-events.rules?

Done 

> So we may not be setting events?

There were no events so far, added with the rules

File: rust/src/quic/frames.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/quic/frames.rs:220]: when `parse_tls_extensions` fails, that is on malformed messages...
What are you thinking of ?

**Comment by victorjulien** [comment]
> > Can we also log both the hash and the string link in tls?
> 
> Done in next version
> 
> I did not do the detection part, as it is only applying a to_md5 transform...

I'd like to reconsider the `quic.ja3` keyword completely, and consider adding support for the existing `ja3*` keywords. These are not limited to tls by their name, so doing something like `alert quic ... ja3.string; content:"abc";` seems to make most sense to me.

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/7144

**Pull request closed without merging**